src/main/java/io/supertokens/storage/mysql/queries/PasswordlessQueries.java

```diff
@@ -0,0 +1,539 @@
+            pst.setString(1, deviceIdHash);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void deleteDevice_Transaction(Start start, Connection con, String deviceIdHash) throws SQLException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getPasswordlessDevicesTable()
+                + " WHERE device_id_hash = ?";
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, deviceIdHash);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void deleteDevicesByPhoneNumber_Transaction(Start start, Connection con, @Nonnull String phoneNumber)
+            throws SQLException {
+
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getPasswordlessDevicesTable()
+                + " WHERE phone_number = ?";
+
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, phoneNumber);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void deleteDevicesByEmail_Transaction(Start start, Connection con, @Nonnull String email)
+            throws SQLException, StorageQueryException {
+
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getPasswordlessDevicesTable() + " WHERE email = ?";
+
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, email);
+            pst.executeUpdate();
+        }
+    }
+
+    private static void createCode_Transaction(Start start, Connection con, PasswordlessCode code) throws SQLException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getPasswordlessCodesTable()
+                + "(code_id, device_id_hash, link_code_hash, created_at)" + " VALUES(?, ?, ?, ?)";
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, code.id);
+            pst.setString(2, code.deviceIdHash);
+            pst.setString(3, code.linkCodeHash);
+            pst.setLong(4, code.createdAt);
+            pst.executeUpdate();
+        }
+    }
+
+    public static void createCode(Start start, PasswordlessCode code)
+            throws StorageTransactionLogicException, StorageQueryException {
+        start.startTransaction(con -> {
+            Connection sqlCon = (Connection) con.getConnection();
+
+            try {
+                PasswordlessQueries.createCode_Transaction(start, sqlCon, code);
+                sqlCon.commit();
+            } catch (SQLException e) {
+
+                String message = e.getMessage();
+                if (message.contains(
+                        "Cannot add or update a child row: a foreign key constraint fails (`supertokens`.`passwordless_codes`, CONSTRAINT `passwordless_codes_ibfk_1` FOREIGN KEY (`device_id_hash`) REFERENCES `passwordless_devices` (`device_id_hash`) ON DELETE CASCADE ON UPDATE CASCA)")) {
+                    throw new StorageTransactionLogicException(new UnknownDeviceIdHash());
+                }
```

Comment:
You are already checking this in Start.java.. no need to check it here.

Reply:
Actually it makes more sense to add the check here instead of start.java, because in-memory has been done that way. Making this change for consistency.

if ( message.contains("foreign key constraint fails") &&
                        message.contains("passwordless_codes") &&
                        message.contains("device_id_hash") )
                {
                    throw new StorageTransactionLogicException(new UnknownDeviceIdHash());
                }

Reply:
Okay.